src/core/network.js

```diff
@@ -15,88 +15,89 @@ class Network extends EventHandler {
-            user.send('self', {
-                userId: user.id,
-                data: user.data,
-                templates: templates.toData()
-            });
-
-            socket.on('room:create', async (levelId, roomType, callback) => {
+        const handlers = {
+            'room:create': async ({ levelId, roomType }, user) => {
                 try {
                     const room = await this.createRoom(levelId, roomType);
-                    this.joinRoom(room.id, user, callback);
+                    const success = this.joinRoom(room.id, user);
+                    return { success };
                 } catch(ex) {
-                    if (callback) callback({ success: false });
+                    console.log(ex);
+                    return { success: false };
                 }
-            });
-
-            socket.on('room:join', (roomId, callback) => {
-                this.joinRoom(roomId, user, callback);
-            });
-
-            socket.on('room:leave', async (roomId, callback) => {
+            },
+            'room:join': (roomId, user) => {
+                this.joinRoom(roomId, user);
+            },
+            'room:leave': async (roomId, user) => {
                 const room = this.rooms.get(roomId);
 
-                if (!room || !socket.rooms.has(roomId))
+                if (!room || !user.rooms.has(roomId))
                     return;
 
-                socket.leave(roomId);
                 room.leave(user);
 
                 this.fire('room:left', room, user);
-
-                if (callback) callback({ success: true });
-            });
-
-            socket.on('level:save', async (level, callback) => {
+            },
+            'level:save': async (level) => {
                 try {
                     await levels.save(level.scene, level);
-                    if (callback) callback({ success: true });
+                    return { success: true };
                 } catch(ex) {
                     console.log(`unable to save level`);
                     console.error(ex);
-                    if (callback) callback({ success: false });
+                    return { success: true };
                 }
-            });
+            }
+        }
 
-            this.fire('user:connected', user);
+        global.LevelProvider = levelProvider;
+
+        this.server = createServer();
+
+        const self = this;
+        this.server.on('upgrade', function(request, socket, body) {
+            if (WebSocket.isWebSocket(request)) {
+                const ws = new WebSocket(request, socket, body);
+                const user = ws.user = new User(ws, Date.now());
+
+                ws.on('open', (e) => {
+                    user.send('self', {
+                        userId: user.id,
+                        data: user.data,
+                        templates: templates.toData()
+                    });
+
+                    self.fire('user:connected', user);
```

Comment:
Get rid off `ed` in all events. Use before/after if need explicit pre/post events/.